opm/simulators/linalg/bda/BILU0.cpp

```diff
@@ -129,6 +153,233 @@ namespace bda
         return true;
     } // end init()
 
+    // implements Fine-Grained Parallel ILU algorithm (FGPILU), Chow and Patel 2015
+    template <unsigned int block_size>
+    void BILU0<block_size>::chow_patel_decomposition()
+    {
+        const unsigned int bs = block_size;
+        int num_sweeps = 6;
+
+        // split matrix into L and U
+        // also convert U into BSC format (Ut)
+        // Ut stores diagonal for now
+        int num_blocks_L = 0;
+
+        // Ut is actually BSC format
+        std::unique_ptr<BlockedMatrix<bs> > Ut = std::make_unique<BlockedMatrix<bs> >(Umat->Nb, Umat->nnzbs + Umat->Nb);
+
+        Lmat->rowPointers[0] = 0;
+        for (int i = 0; i < Nb+1; i++) {
+            Ut->rowPointers[i] = 0;
+        }
+
+        // for every row
+        for (int i = 0; i < Nb; i++) {
+            int iRowStart = LUmat->rowPointers[i];
+            int iRowEnd = LUmat->rowPointers[i + 1];
+            // for every block in this row
+            for (int ij = iRowStart; ij < iRowEnd; ij++) {
+                int j = LUmat->colIndices[ij];
+                if (i <= j) {
+                    Ut->rowPointers[j+1]++;   // actually colPointers
+                } else {
+                    Lmat->colIndices[num_blocks_L] = j;
+                    memcpy(Lmat->nnzValues + num_blocks_L * bs * bs, LUmat->nnzValues + ij * bs * bs, sizeof(double) * bs * bs);
+                    num_blocks_L++;
+                }
+            }
+            Lmat->rowPointers[i+1] = num_blocks_L;
+        }
+
+        // prefix sum
+        int sum = 0;
+        for (int i = 1; i < Nb+1; i++) {
+            sum += Ut->rowPointers[i];
+            Ut->rowPointers[i] = sum;
+        }
+
+        // for every row
+        for (int i = 0; i < Nb; i++) {
+            int iRowStart = LUmat->rowPointers[i];
+            int iRowEnd = LUmat->rowPointers[i + 1];
+            // for every block in this row
+            for (int ij = iRowStart; ij < iRowEnd; ij++) {
+                int j = LUmat->colIndices[ij];
+                if (i <= j){
+                    int idx = Ut->rowPointers[j]++;
+                    Ut->colIndices[idx] = i;     // actually rowIndices
+                    memcpy(Ut->nnzValues + idx * bs * bs, LUmat->nnzValues + ij * bs * bs, sizeof(double) * bs * bs);
+                }
```

Comment:
Before the loop `{0,1,3,6}` should hold.
The first row gets processed: the condition holds for all 3 blocks (all belong in Ut), so the first 3 Ut_rows[] are increased. Now the values are `{1,2,4,6}`.
The second row has 2 blocks that belong in Ut, so the new values are `{1,3,5,6}`.
After the third row it's `{1,3,6,6}`. Now shifting all values to the right (adding a 0 at the start) restores the values.
The block-copying cannot be combined into 1 big copy per row, since Ut is translated. A counter like `num_blocks_L` can also not be used, every row/col would need a separate counter. Ut_rows acts as that counter but needs a reset.

Reply:
After looking at this for a long time, I finally got this code (probably not the smartest person in the project). Please add clear documentation of what you are doing, otherwise other might  be in the same situation as I was.
```
// rowpointer[i] is (mis)used as the write offset of the current row i
...
// now rowpointer[i+1] is at the same position as rowpointer[i] should have for a crs matrix. reset to to correct expected value
```

That is even useful if only you look at the code. (At least I am usually a lot smarter when writing code than when revisiting it)

Reply:
48ebef7